---
title: 'Your Title Here'
author: "Leeza Sergeeva"
date: "November 29, 2020"
output: pdf_document
csl: bioinformatics.csl
bibliography: references.bib
params:
  gff_file_path: "/data/sars_vcf_analysis/02_genome_reference/sars_refgenome_annotation.gff"
  vcf_dir_path: "data/11_vcf_output_for_R"
  sra_runtable_path: "data/00_sra_runtable/SraRunTable_TX_CA.txt"
  chosen_date: "20201130"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      message = FALSE,
                      error = TRUE)
```

# Background and Overview
This report consists of two parts. First part is a variant analysis on compiled SARS-CoV-2 data obtained from two BioProjects on NCBI website. This was an adaptation of the 
previous SARS-CoV-2 variant analysis [@koyama2020variant]. 

This is a report on SARS-CoV-2, including some variant analysis [@koyama2020variant].



# Methods

## SARS-CoV-2 Sequences

I downloaded SraRunTables from two BioProjects on NCBI website related to SARS-CoV-2. Here are the links to "TX SARS-CoV-2 Sequencing" BioProject "https://www.ncbi.nlm.nih.gov/bioproject/PRJNA639066" and "Genome sequencing of Severe acute respiratory syndrome coronavirus 2 in California" BioProject "https://www.ncbi.nlm.nih.gov/bioproject/PRJNA629889". Both SraRunTables were downloaded on November 16, 2020 and combined together into one file SraRunTables_TX_CA.txt.

Reference SARS-CoV-2 isolate Wuhan-Hu-1, complete genome sequence was downloaded on October 15, 2020 from NCBI website "https://www.ncbi.nlm.nih.gov/nuccore/NC_045512". 

## Sequence data processing and filtering, from .fastq to .vcf

Through series of executed shell scripts the .fastq files referenced in SraRunTables_TX_CA.txt file were downloaded to external server provide by the University of San Francisco, as well as the SARS-CoV-2 reference genome file, the annotation gff for SARS-CoV-2. Then [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) program processed .fastq files downloaded to validate the quality of the high throughput sequencing datasets [@andrews2012]. Next step used [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) function to clean up sequencing data by throwing out bad sequences. With the help of [Burrows-Wheeler Alignment Tool](http://bio-bwa.sourceforge.net/bwa.shtml) reference genome was indexed, and the reads in each of the samples provided were aligned to the reference genome and saved as .sam files [@BWA]. Using [Samtools](https://www.htslib.org/), .sam files were converted to .bam files and sorted by leftmost coordinates. Then each sorted .bam file was processed by [Samtools flagstat](https://www.htslib.org/doc/samtools-flagstat.html) to count number of alignments in each FLAG type, like QC pass, QC fail etc [@SAM]. Using [bcftools mpileup](https://samtools.github.io/bcftools/bcftools.html) on sorted .bam files generate .bcf file that contains genotype likelihoods at each genomic position with coverage and then SNPs are called for each input file saving output of that as .vcf files and filtering out the short variants for the final VCF.  

See the set of tutorials on the [vcfR package website](https://knausb.github.io/vcfR_documentation/index.html).

You may also want to use any of a range of different COVID data packages and data sources:

* https://kjhealy.github.io/covdata/
* https://github.com/como-ph/oxcovid19
* https://ropensci.org/blog/2020/10/20/searching-medrxivr-and-biorxiv-preprint-data/
* https://covidtracking.com/data/api
  * `readr::read_csv("https://api.covidtracking.com/v1/states/daily.csv")`
* https://rt.live/
  * `readr::read_csv("https://d14wlfuexuxgcm.cloudfront.net/covid/rt.csv")`




## Subsections are ok too

# Results and Discussion

```{r load-packages-and-functions}
library("vcfR")
library("ggplot2")
library("dplyr")
library("ggthemes")
library("tidyr")
library("readr")
library("ggpubr")

# this loads all function files in the code/functions directory
function_paths <- list.files("code/functions", full.names = TRUE)
purrr::walk(function_paths, source)
```

```{r load-vcf-data}
# load in, tidy, and stack all of the VCF files in the target directory
# these are the output of the bash pipeline
stacked_vcfs <- parse_tidy_and_stack_vcfs(
  vcf_dir_path = params$vcf_dir_path)

# load in the gff file with genome annotations for SARS-CoV-2
gff <- read_gff(gff_file_path = params$gff_file_path)

# pull out just the named genes from the annotation file
gene_table <- extract_genes_from_gff(annotation_object = gff)

# combine the stacked vcf table with the gene names and the metadata from
# the sra runtable file
vcf_with_metadata <- add_genes_metadata_to_vcfstack(
  sra_runtable_path = params$sra_runtable_path,
  stacked_vcf = stacked_vcfs,
  cleaned_genes_table = gene_table)
```

# Figures

```{r sample-plot}
# create a plot of unique SNP locations within each gene across all samples
vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene) %>%
  tally() %>% # this collapses that down to the number of unique SNP locations
  ggplot(aes(x = gene,
             y = n)) +
    geom_col() +
    labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
         x = "Gene Name") +
  theme_few() # get rid of the grey background
```

**Figure 1**: N and S genes have more unique SNPs in the set of samples analyzed.

# Tables

```{r example-table}
# An example table to show the length of each gene using its start and end
gene_table %>%
  mutate(length = end - start) %>%
  select(gene_name, start, end, length) %>%
  knitr::kable(col.names = c("Gene Name",
                             "Start",
                             "End",
                             "Length"))
```

**Table 1**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

# Total Number of Deaths In the US States
``` {r total-deaths-plot, fig.height=12}
# import new data as .csv from covidtracking.com website using API
covid_tracking_data <- readr::read_csv(params$covid_data_api)
  
# dplyr chains, extract columns needed
state_death_summary <- covid_tracking_data %>%
  select(date, state, death, positive, negative, totalTestResults)
#  tally()
  
# number of total deaths by state
total_state_death <- aggregate(death~state, state_death_summary, sum)
names(total_state_death)[2] <- 'total_death'
  
# reorder data to show the states with the most deaths first
total_state_death_ordered <- total_state_death[
  order(total_state_death$total_death, decreasing=TRUE), ]

# make horizontal bar plot with states and total number of deaths
total_state_death_plot <- ggplot(data = total_state_death_ordered,
                         aes(x = total_death,
                             y = reorder(state, total_death))) +
  geom_col() +
  labs(title = "Total Number of Deaths per State",
       x = "Number of Deaths",
       y = "US State",
       labels=TRUE)
#  geom_text(aes(label=total_death))

ggsave(plot = total_state_death_plot,
       filename = "data/images/total_state_death_plot.png")

total_state_death_plot
```
**Figure 2**: The bar plot of states with the total number of Covid-19 related deaths sorted from highest to lowest.


# Total Number of Deaths In the US States
``` {r total-deaths-plot, fig.height=12}
# import death data using a function state_covid_death with new parameter for date
state_death_chosen_date_ten <- state_covid_death(params$chosen_date)
  
total_state_death_plot <- ggplot(data = state_death_chosen_date_ten, 
                                 aes(x = death, y = reorder(state, death))) +
  geom_col() +
  labs(title = paste("Total Number of Deaths per State as of ", params$chosen_date),
       x = "Number of Deaths",
       y = "US State",
       labels=TRUE) +
  geom_text(aes(label=death))

ggsave(plot = total_state_death_plot,
       filename = "data/images/total_state_death_plot_2.png")

total_state_death_plot
```
**Figure 2**: Top 10 states in the United States sorted by the total COVID-19 death numbers

# Sources Cited
